File: vkxml2java/codegen/structure.py
from ..entity import Structure, Member, Registry
from .datatype.common_accessor import *
from .datatype.ptr_accessor import generate_pointer_accessor
from .datatype.array_accessor import generate_array_accessor

def generate_structure(registry: Registry, structure: Structure) -> str:
    member_types_lowered: list[CType | None] = []

    struct_layout = generate_structure_layout(
        registry,
        structure,
        structure.members,
        member_types_lowered
    )

    return f'''package tech.icey.vk4j.datatype;

import java.lang.foreign.*;
import static java.lang.foreign.ValueLayout.*;

import tech.icey.vk4j.annotations.*;
import tech.icey.vk4j.array.*;
import tech.icey.vk4j.bitmask.*;
import tech.icey.vk4j.datatype.*;
import tech.icey.vk4j.enumtype.*;
import tech.icey.vk4j.handle.*;
import tech.icey.vk4j.ptr.*;
import tech.icey.vk4j.NativeLayout;
import tech.icey.vk4j.IDataTypeFactory;
import static tech.icey.vk4j.Constants.*;
import static tech.icey.vk4j.enumtype.VkStructureType.*;

public record {structure.name}(MemorySegment segment) {{
    public static final MemoryLayout LAYOUT = {struct_layout};

{generate_structure_path_element(structure.members)}
{generate_structure_member_layout(structure.members, member_types_lowered)}
{generate_structure_member_offset(structure.members)}
    public {structure.name}(MemorySegment segment) {{
        this.segment = segment;{generate_member_init(structure.members)}
    }}

{generate_structure_member_accessor(structure.members, member_types_lowered)}
{generate_structure_factory(structure)}
}}
'''

def generate_structure_layout(
        registry: Registry,
        structure: Structure,
        members: list[Member],
        member_types_lowered: list[CType | None]
) -> str:
    field_layouts = []

    i = 0
    while i < len(members):
        current = members[i]
        if current.bits is not None:
            assert i + 1 < len(members)

            next_ = members[i + 1]
            assert current.bits is not None and current.bits == 24 and next_.bits == 8

            field_layouts.append(f'ValueLayout.JAVA_INT.withName("bitfield${current.name}_{next_.name}")')
            i = i + 2

            member_types_lowered.append(None)
            member_types_lowered.append(None)
        else:
            ctype = lower_type(registry, current.type)
            field_layouts.append(f'{ctype.java_layout()}.withName("{current.name}")')
            i = i + 1

            member_types_lowered.append(ctype)

    layout_creator = 'NativeLayout.unionLayout' if structure.is_union else 'NativeLayout.structLayout'

    ret = f'{layout_creator}(\n'
    for (i, layout) in enumerate(field_layouts):
        if i == len(field_layouts) - 1:
            ret += f'        {layout}\n'
        else:
            ret += f'        {layout},\n'
    ret += '    )'
    return ret


def generate_structure_path_element(members: list[Member]) -> str:
    ret = ''
    group_element_idx = 0

    i = 0
    while i < len(members):
        current = members[i]
        if current.bits is not None:
            assert i + 1 < len(members)

            next_ = members[i + 1]
            assert current.bits is not None and current.bits == 24 and next_.bits == 8

            ret += f'    public static final PathElement PATH$bitfield${current.name}_{next_.name} = PathElement.groupElement({group_element_idx});\n'
            i = i + 2
        else:
            ret += f'    public static final PathElement PATH${current.name} = PathElement.groupElement({group_element_idx});\n'
            i = i + 1
        group_element_idx += 1

    return ret


def generate_structure_member_layout(members: list[Member], member_types_lowered: list[CType | None]) -> str:
    ret = ''

    i = 0
    while i < len(members):
        current = members[i]
        if current.bits is not None:
            assert i + 1 < len(members)

            next_ = members[i + 1]
            assert current.bits is not None and current.bits == 24 and next_.bits == 8

            ret += f'    public static final OfInt LAYOUT$bitfield${current.name}_{next_.name} = (OfInt) LAYOUT.select(PATH$bitfield${current.name}_{next_.name});\n'
            i += 2
        else:
            ctype = member_types_lowered[i]
            if isinstance(ctype, CPlatformDependentIntType):
                pass
            else:
                ret += f'    public static final {ctype.java_layout_type()} LAYOUT${current.name} = ({ctype.java_layout_type()}) LAYOUT.select(PATH${current.name});\n'
            i += 1

    return ret


def generate_structure_member_offset(members: list[Member]) -> str:
    ret = ''

    i = 0
    while i < len(members):
        current = members[i]
        if current.bits is not None:
            assert i + 1 < len(members)

            next_ = members[i + 1]
            assert current.bits is not None and current.bits == 24 and next_.bits == 8

            ret += f'    public static final long OFFSET$bitfield${current.name}_{next_.name} = LAYOUT.byteOffset(PATH$bitfield${current.name}_{next_.name});\n'
            i += 2
        else:
            ret += f'    public static final long OFFSET${current.name} = LAYOUT.byteOffset(PATH${current.name});\n'
            i += 1

    return ret


def generate_member_init(members: list[Member]) -> str:
    ret = ''

    for member in members:
        if member.values is not None:
            ret += f'\n        this.{member.name}({member.values});'

    return ret


def generate_structure_member_accessor(members: list[Member], member_types_lowered: list[CType | None]) -> str:
    ret = ''

    i = 0
    while i < len(members):
        current = members[i]
        if current.bits is not None:
            assert i + 1 < len(members)

            next_ = members[i + 1]
            assert current.bits is not None and current.bits == 24 and next_.bits == 8

            ret += generate_bitfield_accessor(current, next_)
            i += 2
        else:
            ctype = member_types_lowered[i]
            if isinstance(ctype, CPlatformDependentIntType):
                ret += generate_platform_dependent_int_accessor(current)
            elif isinstance(ctype, CStructType) or isinstance(ctype, CUnionType) or isinstance(ctype, CHandleType):
                ret += generate_ref_type_accessor(ctype, current)
            elif isinstance(ctype, CFixedIntType) or isinstance(ctype, CFloatType):
                ret += generate_fixed_type_accessor(ctype, current)
            elif isinstance(ctype, CEnumType):
                ret += generate_enum_accessor(ctype, current)
            elif isinstance(ctype, CPointerType):
                ret += generate_pointer_accessor(ctype, current)
            elif isinstance(ctype, CArrayType):
                ret += generate_array_accessor(ctype, current)
            else:
                raise ValueError(f'Unsupported type: {ctype}')
            i += 1

    return ret


def generate_structure_factory(struct: Structure) -> str:
    return f'''    public static final class {struct.name}Factory implements IDataTypeFactory<{struct.name}> {{
        @Override
        public Class<{struct.name}> clazz() {{
            return {struct.name}.class;
        }} 

        @Override
        public MemoryLayout layout() {{
            return {struct.name}.LAYOUT;
        }}

        @Override
        public {struct.name} create(MemorySegment segment) {{
            return createUninit(segment);
        }}
        
        @Override
        public {struct.name} createUninit(MemorySegment segment) {{
            return new {struct.name}(segment);
        }}
    }}

    public static final {struct.name}Factory FACTORY = new {struct.name}Factory();'''

File: codegen-v2/src/main/kotlin/club/doki7/babel/codegen/structure.kt
package club.doki7.babel.codegen

import club.doki7.babel.codegen.accessor.*
import club.doki7.babel.ctype.*
import club.doki7.babel.ctype.lowerType
import club.doki7.babel.registry.ArrayType
import club.doki7.babel.registry.IdentifierType
import club.doki7.babel.registry.PointerType
import club.doki7.babel.registry.RegistryBase
import club.doki7.babel.registry.Structure
import club.doki7.babel.registry.Type
import club.doki7.babel.util.Doc
import club.doki7.babel.util.buildDoc

/**
 * @param name the name of the layout, such as "pNext"
 */
sealed class LayoutField(
    val jType: String?,
    val name: String,
    val value: String,
) {
    val layoutName: String = "LAYOUT$$name"
    val sizeName: String = "SIZE$$name"
    val offsetName: String = "OFFSET$$name"
    abstract val pathName: String

    class Typed(jType: String?, name: String, value: String, val type: CType) :
        LayoutField(jType, name, value) {
        override val pathName: String = "PATH$$name"
        val rawName = "${name}Raw"
    }

    /**
     * The [name] is guaranteed to be `[bitfields.first().bitfieldName]_[bitfields.last().bitfieldName]`
     */
    class Bitfields(jType: String, name: String, value: String, val bitfields: List<Bitfield>, val length: Int) :
        LayoutField(jType, name, value) {
        val bitfieldName: String = "bitfield$$name"
        override val pathName: String = "PATH$$bitfieldName"
    }

    data class Bitfield(val bitfieldName: String, val offset: Int) {
        val bitfieldOffsetName: String = "OFFSET$$bitfieldName"
    }
}

fun generateStructure(
    registryBase: RegistryBase,
    structure: Structure,
    isUnion: Boolean,
    codegenOptions: CodegenOptions,
): Doc = buildDoc {
    val packageName = codegenOptions.packageName
    val className = structure.name.toString()
    val originalTypeName = structure.name.original
    val layouts = mutableListOf<LayoutField>()

    lowerMemberTypes(
        registryBase,
        codegenOptions,
        structure,
        layouts,
    )

    /// region import
    +"package $packageName.datatype;"
    +""
    imports("java.lang.foreign.*")
    imports("java.lang.foreign.ValueLayout.*", true)
    +""
    imports("org.jetbrains.annotations.Nullable")
    imports("org.jetbrains.annotations.NotNull")
    imports("club.doki7.ffm.IPointer")
    imports("club.doki7.ffm.NativeLayout")
    if (layouts.any { it is LayoutField.Bitfields }) {
        imports("club.doki7.ffm.bits.BitfieldUtil")
    }
    imports("club.doki7.ffm.annotation.*")
    imports("club.doki7.ffm.ptr.*")

    if (registryBase.bitmasks.isNotEmpty()) {
        imports("$packageName.bitmask.*")
    }

    if (registryBase.opaqueHandleTypedefs.isNotEmpty()) {
        imports("$packageName.handle.*")
    }

    if (registryBase.enumerations.isNotEmpty()) {
        imports("$packageName.enumtype.*")
    }

    if (registryBase.constants.isNotEmpty()) {
        imports("$packageName.${codegenOptions.constantClassName}.*", true)
    }

    for (extra in codegenOptions.extraImport) {
        imports(extra)
    }

    +""
    /// endregion import

    val seeLink = codegenOptions.seeLinkProvider(structure)

    if (seeLink != null) {
        +"/// Represents a pointer to a $seeLink structure in native memory."
    } else {
        +"/// Represents a pointer to a {@code $originalTypeName} structure in native memory."
    }
    +"///"

    +"/// ## Structure"
    +"///"
    +"/// {@snippet lang=c :"
    +"/// typedef struct $originalTypeName {"
    structure.members.forEachIndexed { idx, it ->
        val fieldLink = " @link substring=\"${it.name}\" target=\"#${it.name}\""

        if (it.bits != null) {
            if (it.name.original.startsWith("reserved")) {
                +"///     ${it.type.cDisplay} ${it.name} : ${it.bits};"
            } else {
                +"///     ${it.type.cDisplay} ${it.name} : ${it.bits}; //$fieldLink"
            }
        } else {
            val maybeOptionalComment = if (it.optional) " // optional" else ""

            val cType = (layouts[idx] as LayoutField.Typed).type
            val fieldTypeSubstring = tryFindRootNonPlainType(cType)
            val fieldTypeTarget = tryFindIdentifierType(it.type)
            val maybeTypeLink = if (fieldTypeSubstring != null && fieldTypeTarget != null) {
                " @link substring=\"$fieldTypeSubstring\" target=\"${fieldTypeTarget}\""
            } else {
                ""
            }
            +"///     ${it.type.cDisplay} ${it.name};$maybeOptionalComment //$maybeTypeLink$fieldLink"
        }
    }
    +"/// } $originalTypeName;"
    +"/// }"
    +"///"

    val autoInitMembers = structure.members.filter { it.values != null && it.type is IdentifierType }
    if (autoInitMembers.isNotEmpty()) {
        +"/// ## Auto initialization"
        +"/// This structure has the following members that can be automatically initialized:"
        autoInitMembers.forEach {
            +"/// - `${it.name} = ${it.values!!.original}`"
        }
        +"///"
        +"/// The {@link $className#allocate} functions will automatically initialize these fields."
        +"/// Also, you may call {@link $className#autoInit} to initialize these fields manually for"
        +"/// non-allocated instances."
        +"///"
    }

    +"/// ## Contracts"
    +"///"
    +"/// The property {@link #segment()} should always be not-null"
    +"/// (({@code segment != NULL && !segment.equals(MemorySegment.NULL)}), and properly aligned to)"
    +"/// {@code LAYOUT.byteAlignment()} bytes. To represent null pointer, you may use a Java"
    +"/// {@code null} instead. See the documentation of {@link IPointer#segment()} for more details."
    +"///"
    +"/// The constructor of this class is marked as {@link UnsafeConstructor}, because it does not"
    +"/// perform any runtime check. The constructor can be useful for automatic code generators."

    if (seeLink != null) {
        +"///"
        +"/// @see $seeLink"
    }

    +"@ValueBasedCandidate"
    +"@UnsafeConstructor"
    +"public record $className(@NotNull MemorySegment segment) implements IPointer {"

    indent {
        defun("public static", className, "allocate", "Arena arena") {
            +"$className ret = new $className(arena.allocate(LAYOUT));"
            if (autoInitMembers.isNotEmpty()) {
                if (autoInitMembers.size == 1) {
                    val it = autoInitMembers.first()
                    +"ret.${it.name}(${(it.type as IdentifierType).ident}.${it.values});"
                } else {
                    +"ret.autoInit();"
                }
            }
            +"return ret;"
        }
        +""

        defun("public static", "$className[]", "allocate", "Arena arena", "int count") {
            +"MemorySegment segment = arena.allocate(LAYOUT, count);"
            +"$className[] ret = new $className[count];"
            "for (int i = 0; i < count; i ++)" {
                +"ret[i] = new $className(segment.asSlice(i * BYTES, BYTES));"
                if (autoInitMembers.isNotEmpty()) {
                    if (autoInitMembers.size == 1) {
                        val it = autoInitMembers.first()
                        +"ret[i].${it.name}(${(it.type as IdentifierType).ident}.${it.values});"
                    } else {
                        +"ret[i].autoInit();"
                    }
                }
            }

            +"return ret;"
        }
        +""

        defun("public static", className, "clone", "Arena arena", "$className src") {
            +"$className ret = allocate(arena);"
            +"ret.segment.copyFrom(src.segment);"
            +"return ret;"
        }
        +""

        defun("public static", "$className[]", "clone", "Arena arena", "$className[] src") {
            +"$className[] ret = allocate(arena, src.length);"
            "for (int i = 0; i < src.length; i ++)" {
                +"ret[i].segment.copyFrom(src[i].segment);"
            }
            +"return ret;"
        }
        +""

        if (autoInitMembers.isNotEmpty()) {
            +"public void autoInit() {"
            indent {
                autoInitMembers.forEach {
                    +"${it.name}(${(it.type as IdentifierType).ident}.${it.values});"
                }
            }
            +"}"
            +""
        }

        layouts.forEach {
            when (it) {
                is LayoutField.Bitfields -> +generateBitfieldAccessor(it)
                is LayoutField.Typed -> +generateMemberAccessor(it)
            }
            +""
        }

        // layout of the whole structure (or union)
        val layoutType = if (isUnion) "UnionLayout" else "StructLayout"
        val layoutCtor = if (isUnion) "NativeLayout.unionLayout" else "NativeLayout.structLayout"

        if (layouts.isEmpty()) {
            +"public static final $layoutType LAYOUT = $layoutCtor();"
        } else {
            +"public static final $layoutType LAYOUT = $layoutCtor("
            indent {
                layouts.forEachIndexed { idx, layout ->
                    if (idx == layouts.size - 1) {
                        +layout.value
                    } else {
                        +"${layout.value},"
                    }
                }
            }
            +");"
        }
        defConst("long", "BYTES", "LAYOUT.byteSize()")
        +""

        // `PathElement`s
        layouts.forEach { layout ->
            defConst("PathElement", layout.pathName, "PathElement.groupElement(\"${layout.pathName}\")")
        }

        +""

        // separate layouts
        layouts.forEach { layout ->
            if (layout.jType != null) {
                defConst(layout.jType, layout.layoutName, "(${layout.jType}) LAYOUT.select(${layout.pathName})")
            }
        }

        +""

        // size
        layouts.forEach { layout ->
            if (layout is LayoutField.Typed) {
                val value = if (layout.type is CPlatformDependentIntType) {
                    if (layout.type.cType == "size_t") "NativeLayout.C_SIZE_T.byteSize()"
                    else if (layout.type.cType == "long") "NativeLayout.C_LONG.byteSize()"
                    else if (layout.type.cType == "int") "Integer.BYTES"
                    else error("Unsupported type ${layout.type.cType}")
                } else "${layout.layoutName}.byteSize()"

                defConst("long", layout.sizeName, value)
            }
        }

        +""

        // offset
        layouts.forEach { layout ->
            defConst("long", layout.offsetName, "LAYOUT.byteOffset(${layout.pathName})")
        }
    }

    +"}"
}

private fun generateStructureLayout(layouts: List<LayoutField>, isUnion: Boolean): String {
    return buildString {
        val ctor = if (isUnion) "NativeLayout.unionLayout" else "NativeLayout.structLayout"
        append(ctor)
        layouts.joinTo(this, prefix = "(", postfix = ")") { it.layoutName }
    }
}

private fun lowerMemberTypes(
    registry: RegistryBase,
    codegenOptions: CodegenOptions,
    structure: Structure,
    layouts: MutableList<LayoutField>,
) {
    var i = 0
    while (i < structure.members.size) {
        val current = structure.members[i]
        val cType = lowerType(registry, codegenOptions.refRegistries, current.type)
        if (current.bits != null) {
            val storageUnitBits = (cType as CFixedSizeType).byteSize * 8
            var storageUnitBegin = i
            var alreadyStoredBits: Int = current.bits
            i += 1

            var storageUnitCounter = 0

            while (true) {
                if (i >= structure.members.size) {
                    break
                }

                val field = structure.members[i]
                if (field.bits == null) {
                    break
                }

                if (field.type != current.type) {
                    val fieldTypeLowered = lowerType(
                        registry,
                        codegenOptions.refRegistries,
                        field.type
                    )

                    if ((fieldTypeLowered as CFixedSizeType).byteSize != cType.byteSize) {
                        // Quit the loop, summarize the last storage unit. The new bitfield
                        // would be handled in the next iteration of the outer loop
                        break
                    }
                }

                if (field.bits > storageUnitBits) {
                    error("Field ${field.name} exceeds storage unit size ($storageUnitBits)")
                }

                if (alreadyStoredBits + field.bits > storageUnitBits) {
                    // summarize previous bit fields
                    val storageUnitEnd = i - 1
                    summarizeBitfieldStorageUnit(
                        structure,
                        cType,
                        storageUnitBegin,
                        storageUnitEnd,
                        layouts
                    )

                    // begin a new storage unit
                    storageUnitBegin = i
                    alreadyStoredBits = 0
                    storageUnitCounter += 1
                }

                alreadyStoredBits += field.bits
                i += 1
            }

            // summarize the last storage unit
            val storageUnitEnd = i - 1
            summarizeBitfieldStorageUnit(
                structure,
                cType,
                storageUnitBegin,
                storageUnitEnd,
                layouts
            )
        } else {
            val layout = "${cType.jLayout}.withName(\"${current.name}\")"

            val layoutType = if (cType is CPlatformDependentIntType) {
                null
            } else {
                cType.jLayoutType
            }

            layouts.add(LayoutField.Typed(layoutType, current.name.toString(), layout, cType))

            i += 1
        }
    }
}

private fun summarizeBitfieldStorageUnit(
    structure: Structure,
    storageUnitType: CType,
    storageUnitBegin: Int,
    storageUnitEnd: Int,
    layouts: MutableList<LayoutField>,
) {
    val storageUnitBeginMemberName = structure.members[storageUnitBegin].name.value
    val storageUnitEndMemberName = structure.members[storageUnitEnd].name.value

    val storageUnitName = "${storageUnitBeginMemberName}_${storageUnitEndMemberName}"
    val layout = "${storageUnitType.jLayout}.withName(\"bitfield$$storageUnitName\")"

    val bitfields = mutableListOf<LayoutField.Bitfield>()
    var bitfieldOffset = 0
    for (i in storageUnitBegin..storageUnitEnd) {
        val member = structure.members[i]
        bitfields.add(LayoutField.Bitfield(member.name.toString(), bitfieldOffset))
        bitfieldOffset += member.bits!!
    }

    layouts.add(
        LayoutField.Bitfields(
            storageUnitType.jLayoutType, storageUnitName, layout,
            bitfields, bitfieldOffset
        )
    )
}

private fun generateMemberAccessor(layout: LayoutField.Typed): Doc {
    return when (val cType = layout.type) {
        is CArrayType -> generateArrayAccessor(cType, layout)
        is CHandleType -> generateHandleAccessor(cType, layout)
        is CNonRefType -> generateNonRefAccessor(cType, layout)
        is CPointerType -> generatePtrAccessor(cType, layout)
        is CStructType -> generateStructureTypeAccessor(cType, layout)
        is CVoidType -> throw Exception("void type not allowed in struct")
    }
}

private fun tryFindRootNonPlainType(cType: CType): String? = when (cType) {
    is CArrayType -> tryFindRootNonPlainType(cType.element)
    is CPointerType -> tryFindRootNonPlainType(cType.pointee)
    is CHandleType -> cType.name
    is CEnumType -> cType.name
    is CStructType -> cType.name
    is CVoidType, is CFixedIntType, is CFloatType, is CPlatformDependentIntType -> null
}

private fun tryFindIdentifierType(type: Type): String? = when (type) {
    is ArrayType -> tryFindIdentifierType(type.element)
    is PointerType -> tryFindIdentifierType(type.pointee)
    is IdentifierType -> type.ident.toString()
}
